### PR TITLE
Hood sim

### PR DESCRIPTION
**Pull request by rakrakon**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/hood/HoodIOSim.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                        .withFeedForward(
+                                motorFeedforward.calculate(
+                                        angle.in(Units.RotationsPerSecond.getUnit()))));
+    }
+
+    @Override
+    public void setPower(double power) {
+        inputs.controlMode = Mode.POWER;
+        inputs.powerSetpoint = power;
+        motor.setControl(new VelocityVoltage(power));
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/hood/HoodIOSim.java:49]: When setting simple power, use the `DutyCycleOut` request. The `VelocityVoltage` uses PID to control the motor. Also, the request shouldn't be created every loop, it should be like the position control request.

File: src/main/java/frc/robot/subsystems/hood/HoodIOSim.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import edu.wpi.first.units.Angle;
+import edu.wpi.first.units.MutableMeasure;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.wpilibj.Timer;
+import lib.motors.TalonFXSim;
+
+public class HoodIOSim implements HoodIO {
+    private final TalonFXSim motor;
+
+    private final PositionVoltage control = new PositionVoltage(0);
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/hood/HoodIOSim.java:16]: Should be motion magic.

**Review by vichik123 — CHANGES_REQUESTED** [review]
Good job!

File: src/main/java/frc/robot/subsystems/hood/HoodIOSim.java (hunk 2 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/HoodIOSim.java:16]: e820983

File: src/main/java/frc/robot/subsystems/hood/HoodIOSim.java (hunk 1 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/HoodIOSim.java:49]: 084a961

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java [hunk 3]
```diff
@@ -4,9 +4,40 @@
 import edu.wpi.first.units.Dimensionless;
 import edu.wpi.first.units.Measure;
 import edu.wpi.first.units.Units;
+import frc.robot.Constants;
+import lib.webconstants.LoggedTunableNumber;
 
 public class HoodConstants {
     public static Translation2d MECHANISM_2D_POSE = new Translation2d(1, 1);
     public static double HOOD_LENGTH = 0.4; // m
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/hood/HoodConstants.java:12]: Use measure

File: src/main/java/frc/robot/subsystems/hood/HoodIOSim.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    public void setAngle(MutableMeasure<Angle> angle) {
+        inputs.controlMode = Mode.ANGLE;
+        inputs.angleSetpoint = angle.mut_replace(angle);
+        motor.setControl(
+                control.withPosition(angle.in(Units.Radians))
+                        .withFeedForward(
+                                motorFeedforward.calculate(
+                                        angle.in(Units.RotationsPerSecond.getUnit()))));
+    }
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/hood/HoodIOSim.java:43]: If you have five consecutive parentheses (`)))))`), you should probably extract something into a variable.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java (hunk 3 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/HoodConstants.java:12]: ff20c87

File: src/main/java/frc/robot/subsystems/hood/HoodIOSim.java (hunk 4 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/HoodIOSim.java:43]: ad207d2

**Review by katzuv — CHANGES_REQUESTED** [review]
Sorry, I forgot about that – missing logging the robot-relative position of the shooter. @vichik123 can help if needed.

**Review by vichik123 — DISMISSED** [review]

**Pull request closed without merging**